tawazi/profile.py

```diff
@@ -0,0 +1,79 @@
+from time import perf_counter, process_time, thread_time
+from typing import Any
+
+from tawazi.errors import TawaziTypeError
+
+
+class Profile:
+    def __init__(self, active: bool = True) -> None:
+        """
+        Profile an execution
+
+        Args:
+            active (bool): Whether to activate the profile. Defaults to True.
+
+        >>> p1, p2, = Profile(), Profile()
+        >>> p1 == p2
+        True
+        >>> p2.start()
+        >>> p2.finish()
+        >>> p1 < p2
+        True
+        >>> p1 == p2
+        False
+        >>> p1 > p2
+        False
+        >>> l = sorted([p2, p1])
+        >>> l == [p1, p2]
+        True
+        """
+        self.active = active
+        # [s.] (perf_counter()) system wide time
+        self.abs_exec_time = 0.0
+        # [s.] (process_time()) system & user time consumed by the process
+        self.process_exec_time = 0.0
+        # [s.] (thread_time()) system & user time consumed by the thread
+        self.thread_exec_time = 0.0
+
+    def start(self) -> None:
+        if not self.active:
+            return
+
+        self.abs_exec_time = perf_counter()
+        self.process_exec_time = process_time()
+        self.thread_exec_time = thread_time()
+
+    def finish(self) -> None:
+        if not self.active:
+            return
+        self.abs_exec_time = perf_counter() - self.abs_exec_time
+        self.process_exec_time = process_time() - self.process_exec_time
+        self.thread_exec_time = thread_time() - self.thread_exec_time
+
+    def __enter__(self) -> "Profile":
+        self.start()
```

Comment:
is it really useful to still have a start a finish method in this case ? Can't we merge the __enter__ (resp __exit__) and start (resp finish) method ?

Reply:
` I imagine something like starting the profiling inside an if statement and ending it outside this if statement...` sounds like a context manager use case to me though ahaha